File: mobile/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ApiService {
  static const String baseUrl = 'http://localhost:3000/api';

  // Budget API
  Future<List<dynamic>> getBudgets() async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/budgets'));
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'] ?? [];
      }
      return [];
    } catch (e) {
      print('Error fetching budgets: $e');
      return [];
    }
  }

  Future<Map<String, dynamic>?> getBudgetById(String id) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/budgets/$id'));
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'];
      }
      return null;
    } catch (e) {
      print('Error fetching budget: $e');
      return null;
    }
  }

  Future<List<dynamic>> getBudgetTransactions(String budgetId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/budgets/$budgetId/transactions'),
      );
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'] ?? [];
      }
      return [];
    } catch (e) {
      print('Error fetching transactions: $e');
      return [];
    }
  }

  Future<bool> addTransaction(String budgetId, Map<String, dynamic> transaction) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/budgets/$budgetId/transactions'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(transaction),
      );
      return response.statusCode == 201;
    } catch (e) {
      print('Error adding transaction: $e');
      return false;
    }
  }

  // Portfolio API
  Future<List<dynamic>> getPortfolios() async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/portfolios'));
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'] ?? [];
      }
      return [];
    } catch (e) {
      print('Error fetching portfolios: $e');
      return [];
    }
  }

  Future<Map<String, dynamic>?> getPortfolioById(String id) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/portfolios/$id'));
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'];
      }
      return null;
    } catch (e) {
      print('Error fetching portfolio: $e');
      return null;
    }
  }

  Future<Map<String, dynamic>?> getPortfolioPerformance(String portfolioId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/portfolios/$portfolioId/performance'),
      );
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'];
      }
      return null;
    } catch (e) {
      print('Error fetching performance: $e');
      return null;
    }
  }

  Future<List<dynamic>> getPortfolioTransactions(String portfolioId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/portfolios/$portfolioId/transactions'),
      );
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'] ?? [];
      }
      return [];
    } catch (e) {
      print('Error fetching transactions: $e');
      return [];
    }
  }

  Future<bool> addPortfolioTransaction(
    String portfolioId,
    Map<String, dynamic> transaction,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/portfolios/$portfolioId/transactions'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(transaction),
      );
      return response.statusCode == 201;
    } catch (e) {
      print('Error adding transaction: $e');
      return false;
    }
  }

  // Stock API
  Future<Map<String, dynamic>?> getStockQuote(String symbol) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/stocks/quote/$symbol'));
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'];
      }
      return null;
    } catch (e) {
      print('Error fetching stock quote: $e');
      return null;
    }
  }

  Future<List<dynamic>> searchStocks(String query) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/stocks/search?q=$query'));
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['data'] ?? [];
      }
      return [];
    } catch (e) {
      print('Error searching stocks: $e');
      return [];
    }
  }
}

File: apps/android/app/src/main/java/com/ovaflus/app/data/remote/ApiService.kt
package com.ovaflus.app.data.remote

import com.ovaflus.app.domain.models.Budget
import com.ovaflus.app.domain.models.Holding
import com.ovaflus.app.domain.models.Portfolio
import com.ovaflus.app.domain.models.StockQuote
import com.ovaflus.app.domain.models.Transaction
import com.ovaflus.app.domain.models.User
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface ApiService {

    // Auth
    @GET("auth/me")
    suspend fun getCurrentUser(): User

    // Budgets
    @GET("budgets")
    suspend fun getBudgets(): List<Budget>

    @GET("budgets/{id}")
    suspend fun getBudgetById(@Path("id") id: String): Budget

    @POST("budgets")
    suspend fun createBudget(@Body budget: Budget): Budget

    @PUT("budgets/{id}")
    suspend fun updateBudget(@Path("id") id: String, @Body budget: Budget): Budget

    @DELETE("budgets/{id}")
    suspend fun deleteBudget(@Path("id") id: String)

    // Transactions
    @GET("transactions")
    suspend fun getTransactions(@Query("budgetId") budgetId: String? = null): List<Transaction>

    @POST("transactions")
    suspend fun createTransaction(@Body transaction: Transaction): Transaction

    @DELETE("transactions/{id}")
    suspend fun deleteTransaction(@Path("id") id: String)

    // Portfolios
    @GET("portfolios")
    suspend fun getPortfolios(): List<Portfolio>

    @GET("portfolios/{id}")
    suspend fun getPortfolioById(@Path("id") id: String): Portfolio

    @GET("portfolios/{id}/holdings")
    suspend fun getHoldings(@Path("id") portfolioId: String): List<Holding>

    // Stocks
    @GET("stocks/{symbol}/quote")
    suspend fun getStockQuote(@Path("symbol") symbol: String): StockQuote

    @GET("stocks/search")
    suspend fun searchStocks(@Query("q") query: String): List<StockQuote>
}
